### PR TITLE
Ilut

### PR DESCRIPTION
**Pull request by robertsj**
Adding a ILUT preconditioner.  With limiting parameters, one can get a full LU (though not robustly).  With sensible parameters, one can get a nice middle ground between ILU(p) and other things.

### CODE REVIEW

File: src/callow/test/test_SparseRow.cc [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+	//              0  1  2  3  4  5  6  7  8
+    double a[10] = {1, 0, 3, 0, 2, 0, 6, 4, 5, };
+    SparseRow row(a, 10);
+    SparseRow row2 = row;
+    row.display("before");
+    std::sort(row.begin(), row.end(), SparseRow::compare);
+    row.display("after");
+
+    row2.display("before 2");
+    std::sort(row2.range(0, 5).first, row2.range(0, 5).second, SparseRow::compare);
```
Comment (Relzohery) [inline src/callow/test/test_SparseRow.cc:76]: looks like compare method is not defined!

**Review by Relzohery — CHANGES_REQUESTED** [review]

File: src/callow/test/test_SparseRow.cc [hunk 2]
```diff
@@ -0,0 +1,178 @@
+//----------------------------------*-C++-*-----------------------------------//
```
Comment (Relzohery) [inline src/callow/test/test_SparseRow.cc:1]: This works for c++11 and above only. Not sure if it matters ...

File: src/callow/matrix/SparseRow.cc [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+  Require(b >= a);
+  auto _begin = begin();
+  for (; _begin < end(); ++_begin)
+  {
+	if (_begin->first > b)
+	  return range_t(end(), end());
+	if (_begin->first >= a)  // insertions are before, so we want next one
+	  break;
+  }
+  if (_begin == cend())
```
Comment (Relzohery) [inline src/callow/matrix/SparseRow.cc:271]:
```suggestion
  if (_begin == end())
```

File: src/callow/test/test_Preconditioners.cc [hunk 4]
```diff
@@ -59,6 +65,156 @@ int test_PCILU0(int argc, char *argv[])
...
+    {
+    	TEST(soft_equiv(got_v[i], expect_v[i]));
+    }
+  }
+
+  return 0;
+}
+
+//----------------------------------------------------------------------------//
+int test_performance(int argc, char *argv[])
```
Comment (Relzohery) [inline src/callow/test/test_Preconditioners.cc:157]: I thought you would check the resultant vector x !

Comment (robertsj) [reply src/callow/test/test_Preconditioners.cc:157]: I'm using right preconditioning, so the residual is still ||A*x_appx-b||. If GMRES converges, it must converge to the right answer.  But worth verifying, for sure.

**Pull request merged**